### DISCUSSION TITLE
Raise window on notification activation?

### DISCUSSION THREAD

**Discussion by winteriscariot** [labwc/labwc#2059] [Q&A]
I've tested this with both mako and swaync, and clicking on a notification sent by an application does not raise that application. have tested with discord, telegram, and signal-desktop

would this be something that would be in labwc or in the notification daemons?

edit: actually I have a sense that this is labwc, since doing vscodium -r file (to open the file in the existing vscodium window) does not raise the window either, and no notification daemons are involved in that

is there an option i'm unaware of to make this behave in such a way?

**Comment by tokyo4j** [comment]
Do you use labwc from the master branch? The master branch includes #1974 which validates xdg-activation token more strictly.

**Reply by winteriscariot** [reply]
i'm currently on 4bda13d which i believe should include that change, yet windows don't raise on activation. is there an rc.xml option i've missed to enable it?

**Reply by tokyo4j** [reply]
#1974 validates xdg-activation token more strictly, meaning labwc is now more likely to ignore activation requests. As we discussed in the PR, we will address (workaround) this issue before the next release.

Could you try 0.7.4 and see if windows are raised on activation?

**Reply by winteriscariot** [reply]
oooooh okay I see, I misunderstood the nature of the PR. i think i understand now. I'll probably just wait til it gets addressed in the future release, as it's not a *huge* inconvenience to me.

thanks!!

**Reply by Consolatis** [reply]
From memory, mako notifications should raise the window even with the PR mentioned when mako is using `on-button-left=invoke-default-action` which is the default configuration. The client also needs to support it though which I am not sure about for the clients mentioned.

@tokyo4j should we get your proposal into 0.8.0? It seems a very minor change. I could write a PR later if you haven't written a prototype already.

**Reply by winteriscariot** [reply]
> mako notifications should raise the window even with the PR mentioned when mako is using `on-button-left=invoke-default-action` which is the default configuration. The client also needs to support it though which I am not sure about for the clients mentioned.

just confirmed it does NOT work with mako+Thunderbird, but to your last point there I have no idea if it *should* work.

**Reply by Consolatis** [reply]
I've looked a bit into this and it seems thunderbird (115.13.0) doesn't use the token it received:
```
00:01:00.137 [../src/input/cursor.c:1036] press on layer-surface
00:01:00.137 [types/wlr_xdg_activation_v1.c:195] Activation token '(null)' seat set
00:01:00.137 [types/wlr_xdg_activation_v1.c:226] Activation token '(null)' surface set
00:01:00.138 [types/wlr_xdg_activation_v1.c:130] New activation token '6c0bfe124bb9a452342dc4c99d405086' created, seat 0xffffeba03080 surface 0xffffee08ca80
00:01:00.204 [types/wlr_xdg_activation_v1.c:195] Activation token '(null)' seat set
00:01:00.204 [types/wlr_xdg_activation_v1.c:130] New activation token '6674e01098d74c1212517c4437d55cf8' created, seat 0xffffeba03080 surface (nil)
00:01:00.207 [../src/xdg.c:771] handling xdg activation request for token 6674e01098d74c1212517c4437d55cf8
00:01:00.207 [../src/xdg.c:792] Denying focus request, source surface not set
00:01:30.138 [types/wlr_xdg_activation_v1.c:46] Activation token '6c0bfe124bb9a452342dc4c99d405086' has expired
```

In that log `6c0bfe124bb9a452342dc4c99d405086` is the valid token requested by mako and `6674e01098d74c1212517c4437d55cf8` a new invalid token (because it didn't have focus) created by thunderbird.

I've also verified that mako returns the correct token:
```
test@odroidc4:/home/user/dev/labwc$ G_MESSAGES_DEBUG=all notify-send --action=default="some action" header body
(notify-send:3558969): libnotify-DEBUG: 18:49:08.240: SNAP app: (null)
(notify-send:3558969): GLib-GIO-DEBUG: 18:49:08.245: Using cross-namespace EXTERNAL authentication (this will deadlock if server is GDBus < 2.73.3)
(notify-send:3558969): libnotify-DEBUG: 18:49:08.257: Using sever mako emersion, v0.0.0 - Supporting Notification Spec 1.2
default
(notify-send:3558969): libnotify-DEBUG: 18:49:11.056: Activation Token: ea27274e74a0322d2f7c9c5a0c12d68b
```
with the corresponding labwc log
```
00:04:30.647 [../src/input/cursor.c:1036] press on layer-surface
00:04:30.648 [types/wlr_xdg_activation_v1.c:195] Activation token '(null)' seat set
00:04:30.648 [types/wlr_xdg_activation_v1.c:226] Activation token '(null)' surface set
00:04:30.648 [types/wlr_xdg_activation_v1.c:130] New activation token 'ea27274e74a0322d2f7c9c5a0c12d68b' created, seat 0xffffeba03080 surface 0xffffee08f780
00:05:00.648 [types/wlr_xdg_activation_v1.c:46] Activation token 'ea27274e74a0322d2f7c9c5a0c12d68b' has expired
```

So this looks like a thunderbird bug as it ignores the valid token received via dbus from the notification daemon and instead creates its own one which now doesn't work due to the stricter checks in labwc.

**Reply by winteriscariot** [reply]
wow thanks for checking into that!! i think for now i'm going to work around the issue with custom mako exec action with wlrctl as that seems pretty straight forward and can consistently raise applications

**Reply by tokyo4j** [reply]
> @tokyo4j should we get your proposal into 0.8.0? It seems a very minor change. I could write a PR later if you haven't written a prototype already.

~Yes, I'd appreciate it if you could write the PR.~

~At least, something like `<focusRequestRequirement input="yes|no" focus="yes|no" />` I suggested in https://github.com/labwc/labwc/pull/1974#issuecomment-2248549710 should be added so we can say to user: "If you want the old behavior, use these settings".~

I opened #2066 that undoes #1974, since for now I'm not sure about how the configuration should look like (e.g. how it should be extended when we implement urgency hint).

**Comment by winteriscariot** [comment]
As a follow up to this, rather than opening a new discussion, I wanted to expand the question: what about raising applications on *any* sort of activation, whether or not it's triggered by a notification?

For example: if I already have pcmanfm-qt open but minimized, and I launch that application again (via menu or rofi) I expect the application to be activated and unminimized. Currently if I run `pcmanfm-qt` while a window is opened and minimized, a new tab is opened but the window stays minimized. Imo this is bad flow: often times I wonder, "wait why isn't pcmanfm launching?" but it _is_ working, it just stays minimized, i just hadn't noticed that the app was already open.

Is there currently a way to change this behavior?
